### PR TITLE
Move ExecuteUpdate/ExecuteDelete to core so non-relational providers can implement

### LINKED ISSUE #31052: Promote ExecuteUpdate/Delete from relational to core
We introduced ExecuteUpdate/Delete at the relational layer. However, there are non-relational database which support filter-based updating, specifically MongoDB ([DeleteMany](https://www.mongodb.com/docs/drivers/csharp/current/usage-examples/deleteMany/), [UpdateMany](https://www.mongodb.com/docs/drivers/csharp/current/usage-examples/updateMany/)), and possibly others.

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #31052

### CODE REVIEW

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs [hunk 1]
```diff
@@ -555,6 +555,13 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
                             thenInclude: false,
                             setLoaded: false);
 
+                    // Handled by relational/provider even though method is on `EntityFrameworkQueryableExtensions`
+                    case nameof(EntityFrameworkQueryableExtensions.ExecuteDelete):
+                    case nameof(EntityFrameworkQueryableExtensions.ExecuteDeleteAsync):
+                    case nameof(EntityFrameworkQueryableExtensions.ExecuteUpdate):
+                    case nameof(EntityFrameworkQueryableExtensions.ExecuteUpdateAsync):
+                        return ProcessUnknownMethod(methodCallExpression);
```
Comment (roji) [inline src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs:563]: We should implement proper nav expansion for ExecuteUpdate (see #32493), but this can be done separately in another PR (this doesn't make things any worse than the current situation, even if it doesn't make them better).

File: test/EFCore.InMemory.FunctionalTests/BulkUpdates/FiltersInheritanceBulkUpdatesInMemoryTest.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.EntityFrameworkCore.BulkUpdates;
+
+public class FiltersInheritanceBulkUpdatesInMemoryTest(FiltersInheritanceBulkUpdatesInMemoryFixture fixture)
```
Comment (roji) [inline test/EFCore.InMemory.FunctionalTests/BulkUpdates/FiltersInheritanceBulkUpdatesInMemoryTest.cs:6]: Does it really make sense to "implement" bulk update test suites for InMemory, given that it's not supported? It just means we'll have to add overrides to InMemory each time we add a test just for translation failure... I'd rather we didn't have to do this (i.e. we can just skip bulk updates tests in the InMemoryComplianceTest)

File: src/EFCore.Relational/Properties/TypeForwards.cs [hunk 3]
```diff
@@ -6,3 +6,4 @@
 [assembly: TypeForwardedTo(typeof(AttributeCodeFragment))]
 [assembly: TypeForwardedTo(typeof(MethodCallCodeFragment))]
 [assembly: TypeForwardedTo(typeof(NestedClosureCodeFragment))]
+[assembly: TypeForwardedTo(typeof(SetPropertyCalls<>))]
```
Comment (roji) [inline src/EFCore.Relational/Properties/TypeForwards.cs:9]: Out of curiosity why is this needed? Does it actually help given that we already have a binary break anyway by moving the ExecuteUpdate/Delete APIs themselves?

**Review by roji — APPROVED** [review]
LGTM, just not sure we should implement the test suites for InMemory.

File: test/EFCore.InMemory.FunctionalTests/BulkUpdates/FiltersInheritanceBulkUpdatesInMemoryTest.cs (hunk 2 above)
Comment (roji) [reply test/EFCore.InMemory.FunctionalTests/BulkUpdates/FiltersInheritanceBulkUpdatesInMemoryTest.cs:6]: Another thought on this... I do agree we should indeed have like a single test just to confirm the behavior for unsupported providers (I'm guessing that was the intention behind implementing all these for InMemory - but I'd just do one somewhere for Update and another for Delete).

Then, for providers which don't support it, we should implement a default translation failure with a specific message saying "your provider does not support ExecuteUpdate/Delete", rather than the generic "translation failed" message. See my comment about QueryableMethodTranslatingExpresionVisitor below for implementing that.

**Comment by roji** [comment]
One more thought... We should follow the pattern of adding TranslateExecuteUpdate/Delete methods to QueryableMethodTranslatingExpresionVisitor, and dispatching to them from the big method dispatch like with the others. Rather than making them abtract like the others, we should just provide a default implementation that adds a specific error message "your provider does not support ExecuteUpdate/Delete" and returns null. Relational would simply override these methods rather than implementing its own special dispatch in VisitMethodCall (which should be removed).

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -226,27 +226,35 @@ when entityQueryRootExpression.GetType() == typeof(EntityQueryRootExpression)
...
+    private static readonly MethodInfo ExecuteUpdateMethodInfo
+        = typeof(EntityFrameworkQueryableExtensions).GetTypeInfo().GetDeclaredMethod(
+            nameof(EntityFrameworkQueryableExtensions.ExecuteUpdate))!;
+
     /// <inheritdoc />
     protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
     {
         var method = methodCallExpression.Method;
-        if (method.DeclaringType == typeof(RelationalQueryableExtensions))
+        if (method.DeclaringType == typeof(EntityFrameworkQueryableExtensions))
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:241]: This dispatch should move up to core QueryableMethodTranslatingEV like the dispatching of the other standard operators.

File: test/EFCore.InMemory.FunctionalTests/BulkUpdates/FiltersInheritanceBulkUpdatesInMemoryTest.cs (hunk 2 above)
Comment (ajcvickers) [reply test/EFCore.InMemory.FunctionalTests/BulkUpdates/FiltersInheritanceBulkUpdatesInMemoryTest.cs:6]: The reason I implemented the in-memory tests was just to give me confidence I had done the refactoring of the relational parts correctly. (I had made several mistakes, btw. It's hard to do an abstraction right without implementing it. Hmm. ;-)) I left them because they were testing the current behavior of non-relational providers, I'm happy to remove them and have a single test with a specific message.

File: src/EFCore.Relational/Properties/TypeForwards.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore.Relational/Properties/TypeForwards.cs:9]: Yeah, I think you're right. But then again, it doesn't do any harm.

**Comment by ajcvickers** [comment]
@roji I moved the dispatch to core.

**Comment by ajcvickers** [comment]
@roji Helps if I actually push the commit.

File: src/EFCore/Query/QueryableMethodTranslatingExpressionVisitor.cs [hunk 5]
```diff
@@ -138,6 +138,32 @@ protected override Expression VisitExtension(Expression extensionExpression)
     protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
     {
         var method = methodCallExpression.Method;
+
+        if (method.DeclaringType == typeof(EntityFrameworkQueryableExtensions))
```
Comment (roji) [inline src/EFCore/Query/QueryableMethodTranslatingExpressionVisitor.cs:142]: I'd just merge this into the block below, this would deduplicate the visitation, check for SQE, etc. (we check on the generic methods for each case anyway)

File: src/EFCore/Query/QueryableMethodTranslatingExpressionVisitor.cs [hunk 6]
```diff
@@ -1000,4 +1026,39 @@ protected virtual Expression MarkShaperNullable(Expression shaperExpression)
...
+    #region ExecuteUpdate/ExecuteDelete
+
+    /// <summary>
+    ///     Translates <see cref="EntityFrameworkQueryableExtensions.ExecuteDelete{TSource}(IQueryable{TSource})" /> method
+    ///     over the given source.
+    /// </summary>
+    /// <param name="source">The shaped query on which the operator is applied.</param>
+    /// <returns>The non query after translation.</returns>
+    protected virtual Expression? TranslateExecuteDelete(ShapedQueryExpression source)
+        => throw new InvalidOperationException(
```
Comment (roji) [inline src/EFCore/Query/QueryableMethodTranslatingExpressionVisitor.cs:1039]: The more standard way to report translation failure is to call AddTranslationErrorDetails() with the error, and then return null (this is e.g. what we do in the relational implementations when we can't translate for some reason). In most cases that's important in order to client eval on translation (which the exception prevents); of course with update/delete there's no client eval, but still a good idea to follow the same practice.

File: test/EFCore.InMemory.FunctionalTests/InMemoryComplianceTest.cs [hunk 7]
```diff
@@ -16,6 +18,12 @@ public class InMemoryComplianceTest : ComplianceTestBase
         typeof(StoreGeneratedTestBase<>),
         typeof(ConferencePlannerTestBase<>),
         typeof(ManyToManyQueryTestBase<>),
+        typeof(ComplexTypeBulkUpdatesTestBase<>),
```
Comment (roji) [inline test/EFCore.InMemory.FunctionalTests/InMemoryComplianceTest.cs:21]: 👍

File: test/EFCore.Relational.Specification.Tests/BulkUpdates/NorthwindBulkUpdatesRelationalTestBase.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    public virtual Task Delete_FromSql_converted_to_subquery(bool async)
+        => TestHelpers.ExecuteWithStrategyInTransactionAsync(
+            () => Fixture.CreateContext(),
+            (facade, transaction) => Fixture.UseTransaction(facade, transaction),
+            async context =>
+            {
+                var queryable = context.Set<OrderDetail>().FromSqlRaw(
+                    NormalizeDelimitersInRawString(
+                        @"SELECT [OrderID], [ProductID], [UnitPrice], [Quantity], [Discount]
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/BulkUpdates/NorthwindBulkUpdatesRelationalTestBase.cs:43]: nit: raw literal string (and below)

File: test/EFCore.Tests/Query/QueryProviderTest.cs [hunk 9]
```diff
@@ -5,6 +5,29 @@ namespace Microsoft.EntityFrameworkCore.Query;
 
 public class QueryProviderTest
 {
+    [ConditionalFact]
+    public async Task ExecuteUpdate_and_ExecuteDelete_throw_when_provider_does_not_implement()
```
Comment (roji) [inline test/EFCore.Tests/Query/QueryProviderTest.cs:9]: 👍

File: src/EFCore/Query/QueryableMethodTranslatingExpressionVisitor.cs (hunk 6 above)
Comment (ajcvickers) [reply src/EFCore/Query/QueryableMethodTranslatingExpressionVisitor.cs:1039]: You get the normal translation failure if the provider supports ExecuteUpdate/ExecuteDelete but the query can't be translated. This is for the case where the provider has not implemented it at all, so there's no client eval concerns. and also I don't think it's helpful to give people the query like they might be able to do something to it to make this work.

In other words, this isn't really a translation failure message, but rather a provider doesn't support this message.

Comment (roji) [reply src/EFCore/Query/QueryableMethodTranslatingExpressionVisitor.cs:1039]: OK, we don't have that distinction anywhere else, but I get the logic... In any case it's all rather academic. Am OK with merging as-is.
